### ISSUE TITLE
Can we use videos or Gif as filter ?

### ISSUE THREAD

**Issue by developerandroid123** [wysaid/android-gpuimage-plus#145]
Can we use video or gif as filter? You have used images hehe.png ,wild etc So can we go with videos ? if yes please help me out this

**Comment by wysaid** [comment]
Sure.
But you should write a new filter with video/gif decoder.
Update& generate the new texture content each frame, then apply to the filter

**Comment by developergurjeet** [comment]
How can i achieve this ? Can you please guide me

**Comment by wysaid** [comment]
Try writing a new filter with 10 images(as a gif), and play them with 30fps.

**Comment by developergurjeet** [comment]
OKAY but it will play at same time and i need to show each image on different frame So it seems like video effect on video
How can i do this?

**Comment by wysaid** [comment]
The same way

**Comment by developergurjeet** [comment]
Can you please give an example for this:
10 images(as a gif), and play them at 30fps.

**Comment by wysaid** [comment]
Do it yourself like [this](https://github.com/wysaid/android-gpuimage-plus/blob/master/library/src/main/jni/source/customFilter_N.cpp)

**Comment by developergurjeet** [comment]
ok let me try

**Comment by developerandroid123** [comment]
If i use @dynamic mf 10 0 then it  gives effect on movement of camera But i want it static movement effect without moving camera
How can i achieve this?

**Comment by wysaid** [comment]
Write a new filter with motion blur

**Comment by developerandroid123** [comment]
motion blur?

**Comment by wysaid** [comment]
[motion blur algorithm](https://www.google.com/search?q=motion+blur+algorithm)

**Comment by developerandroid123** [comment]
I am trying to replace your second custom filter with my filter but its giving me same old filter
where i am doing wrong?
 
CGEConstString s_fsh2 = CGE_SHADER_STRING_PRECISION_H
(
 varying vec2 textureCoordinate;
 uniform sampler2D inputImageTexture;
 precision highp float;
 uniform vec2 vSteps;
 uniform float iGlobalTime;
 uniform float u_Speed;
    void main( )
    {
    	float drunk = sin(iGlobalTime*2.0)*6.0/1024.0;
    	float unitDrunk1 = (sin(iGlobalTime*1.2)+1.0)/2.0;
    	float unitDrunk2 = (sin(iGlobalTime*1.8)+1.0)/2.0;

    	vec2 normalizedCoord = textureCoordinate + vec2(0, drunk);
    	normalizedCoord.x = pow(normalizedCoord.x, mix(1.25, 0.85, unitDrunk1));
    	normalizedCoord.y = pow(normalizedCoord.y, mix(0.85, 1.25, unitDrunk2));

    	vec2 normalizedCoord2 = textureCoordinate + vec2(0, drunk);
    	normalizedCoord2.x = pow(normalizedCoord2.x, mix(0.95, 1.1, unitDrunk2));
    	normalizedCoord2.y = pow(normalizedCoord2.y, mix(1.1, 0.95, unitDrunk1));

    	vec2 normalizedCoord3 = textureCoordinate;

    	vec4 color = texture2D(inputImageTexture, normalizedCoord);
    	vec4 color2 = texture2D(inputImageTexture, normalizedCoord2);
    	vec4 color3 = texture2D(inputImageTexture, normalizedCoord3);

    	// Mess with colors and test swizzling
    	color.x = sqrt(color2.x);
    	color2.x = sqrt(color2.x);

    	vec4 finalColor = mix( mix(color, color2, mix(0.4, 0.6, unitDrunk1)), color3, 0.4);

    	//
    	if (length(finalColor) > 1.4)
    		finalColor.xy = mix(finalColor.xy, normalizedCoord3, 0.5);
    	else if (length(finalColor) < 0.4)
    		finalColor.yz = mix(finalColor.yz, normalizedCoord3, 0.5);

    	gl_FragColor = finalColor;
    }
 );

bool CustomFilter_2::init()
{
    return m_program.initWithShaderStrings(g_vshDefaultWithoutTexCoord, s_fsh2);
}

**Comment by wysaid** [comment]
Change the filter with only red color output, and see what will happen.

**Comment by developerandroid123** [comment]
Which filter do i need to change with red filter?

**Comment by developerandroid123** [comment]
Which filter do i need to change with red filter?

**Comment by wysaid** [comment]
The one which you write.

**Comment by developerandroid123** [comment]
can you please edit my filter and paste here So i can understand easily

**Comment by wysaid** [comment]
```
CGEConstString s_fsh2 = CGE_SHADER_STRING_PRECISION_H
(
void main( )
{
    gl_FragColor = vec4(1.0, 0.0, 0.0, 1.0);
}
```

**Comment by developerandroid123** [comment]
it is doing nothing I dont understand where i am doing wrong

**Comment by wysaid** [comment]
So do I. Did you rebuild the jni part after changing the source?

**Comment by developerandroid123** [comment]
How can i do that?

**Comment by wysaid** [comment]
https://github.com/wysaid/android-gpuimage-plus#abstract

**Comment by developerandroid123** [comment]
android-gpuimage-plus-master/library/src/main/jni/include/cgeImageHandler.h:13:10: fatal error:'CGEImageFilter.h' file not found

This error occurs When i try to build jni with terminal

**Comment by developerandroid123** [comment]
export NDK=path/of/your/ndk
cd folder/of/jni (android-gpuimage-plus/library/src/main/jni)

#This will make all arch: armeabi, armeabi-v7a arm64-v8a, x86
./buildJNI
#Or use "sh buildJNI"

#Try this if you failed to run the shell above
export CGE_USE_VIDEO_MODULE=1
$NDK/ndk-build



I have tried both and giving me same error as i mentioned above

**Comment by wysaid** [comment]
git pull
and try again

**Comment by developerandroid123** [comment]
Same error is coming every time

**Comment by wysaid** [comment]
Clone the repo again.

**Comment by developerandroid123** [comment]
I have clone your repo again and please tell me how to follow further

**Comment by developerandroid123** [comment]
please tell me as i am trying to build changes in jni since morning And its giving me error

**Comment by wysaid** [comment]
see the above, and check your code

**Comment by developerandroid123** [comment]
Yes i have seen that you have included #include "cgeImageFilter.h" 
Do i need to follow these steps same:
export NDK=path/of/your/ndk
cd folder/of/jni (android-gpuimage-plus/library/src/main/jni)

#This will make all arch: armeabi, armeabi-v7a arm64-v8a, x86
./buildJNI
#Or use "sh buildJNI"

#Try this if you failed to run the shell above
export CGE_USE_VIDEO_MODULE=1
$NDK/ndk-build

**Comment by developerandroid123** [comment]
please confirm

**Comment by developerandroid123** [comment]
i have done using same flow after cloning repo again
But unfortunately giving me  error and too bad for me

**Comment by wysaid** [comment]
So what is the error

**Comment by developerandroid123** [comment]
root:~# export NDK=/home/iapp/android-sdk/ndk-bundleroot@iapp-PHP-106:~# cd /home/user/NewAPP/android-gpuimage-plus/library/src/main/jni

root@user-PHP-106:/home/user/NewAPP/android-gpuimage-plus/library/src/main/jni# ./buildJNI

The NDK dir is: /home/user/android-sdk/ndk-bundle, If the shell can not run normally, you should set the NDK variable to your local ndk-dir

[armeabi] Install        : libCGE.so => libs/armeabi/libCGE.so
[armeabi] Compile++ thumb: CGEExt <= customHelper.cpp
In file included from /home/iapp/NewAPP/android-gpuimage-plus/library/src/main/jni/source/customHelper.cpp:11:
/home/iapp/NewAPP/android-gpuimage-plus/library/src/main/jni/source/customFilter_N.h:12:10: fatal error: 
      'CustomFilter_0.h' file not found
#include "CustomFilter_0.h"
         ^
1 error generated.
make: *** [/home/iapp/NewAPP/android-gpuimage-plus/library/src/main/obj/local/armeabi/objs/CGEExt//home/iapp/NewAPP/android-gpuimage-plus/library/src/main/jni/source/customHelper.o] Error 1
root@user:/home/user/NewAPP/android-gpuimage-plus/library/src/main/jni#

**Comment by developerandroid123** [comment]
See this error

**Comment by wysaid** [comment]
pull again

**Comment by zlogicTech** [comment]
that works Thankyou

**Comment by zlogicTech** [comment]
static CGEConstString s_fsh2 = CGE_SHADER_STRING_PRECISION_H
(
 varying vec2 textureCoordinate;
 uniform sampler2D inputImageTexture;
 uniform vec2 vSteps;
 const float intensity = 0.45;
 precision highp float;
 uniform float iGlobalTime;
 uniform float u_Speed;

 void main( )
 {
 	float drunk = sin(iGlobalTime*2.0)*6.0/1024.0;
 	float unitDrunk1 = (sin(iGlobalTime*1.2)+1.0)/2.0;
 	float unitDrunk2 = (sin(iGlobalTime*1.8)+1.0)/2.0;

 	vec2 normalizedCoord = textureCoordinate + vec2(0, drunk);
 	normalizedCoord.x = pow(normalizedCoord.x, mix(1.25, 0.85, unitDrunk1));
 	normalizedCoord.y = pow(normalizedCoord.y, mix(0.85, 1.25, unitDrunk2));

 	vec2 normalizedCoord2 = textureCoordinate + vec2(0, drunk);
 	normalizedCoord2.x = pow(normalizedCoord2.x, mix(0.95, 1.1, unitDrunk2));
 	normalizedCoord2.y = pow(normalizedCoord2.y, mix(1.1, 0.95, unitDrunk1));

 	vec2 normalizedCoord3 = textureCoordinate;

 	vec4 color = texture2D(inputImageTexture, normalizedCoord);
 	vec4 color2 = texture2D(inputImageTexture, normalizedCoord2);
 	vec4 color3 = texture2D(inputImageTexture, normalizedCoord3);

 	// Mess with colors and test swizzling
 	color.x = sqrt(color2.x);
 	color2.x = sqrt(color2.x);

 	vec4 finalColor = mix( mix(color, color2, mix(0.4, 0.6, unitDrunk1)), color3, 0.4);

 	//
 	if (length(finalColor) > 1.4)
 		finalColor.xy = mix(finalColor.xy, normalizedCoord3, 0.5);
 	else if (length(finalColor) < 0.4)
 		finalColor.yz = mix(finalColor.yz, normalizedCoord3, 0.5);

 	gl_FragColor = finalColor;
 }
 );

bool CustomFilter_2::init()
{
        return m_program.initWithShaderStrings(g_vshDefaultWithoutTexCoord, s_fsh2);
}

**Comment by zlogicTech** [comment]
Now i have made this twitch effect But its not working its just give me colored filter
I have used red color output that just give red screen

**Comment by developerandroid123** [comment]
That works Thankyou

**Comment by developerandroid123** [comment]
Now i have made this twitch effect But its not working its just give me colored filter
I have also used red color output that just give red screen

**Comment by developerandroid123** [comment]
static CGEConstString s_fsh2 = CGE_SHADER_STRING_PRECISION_H
(
varying vec2 textureCoordinate;
uniform sampler2D inputImageTexture;
uniform vec2 vSteps;
const float intensity = 0.45;
precision highp float;
uniform float iGlobalTime;
uniform float u_Speed;

void main( )
{
float drunk = sin(iGlobalTime2.0)6.0/1024.0;
float unitDrunk1 = (sin(iGlobalTime1.2)+1.0)/2.0;
float unitDrunk2 = (sin(iGlobalTime1.8)+1.0)/2.0;

vec2 normalizedCoord = textureCoordinate + vec2(0, drunk);
normalizedCoord.x = pow(normalizedCoord.x, mix(1.25, 0.85, unitDrunk1));
normalizedCoord.y = pow(normalizedCoord.y, mix(0.85, 1.25, unitDrunk2));

vec2 normalizedCoord2 = textureCoordinate + vec2(0, drunk);
normalizedCoord2.x = pow(normalizedCoord2.x, mix(0.95, 1.1, unitDrunk2));
normalizedCoord2.y = pow(normalizedCoord2.y, mix(1.1, 0.95, unitDrunk1));

vec2 normalizedCoord3 = textureCoordinate;

vec4 color = texture2D(inputImageTexture, normalizedCoord);
vec4 color2 = texture2D(inputImageTexture, normalizedCoord2);
vec4 color3 = texture2D(inputImageTexture, normalizedCoord3);

// Mess with colors and test swizzling
color.x = sqrt(color2.x);
color2.x = sqrt(color2.x);

vec4 finalColor = mix( mix(color, color2, mix(0.4, 0.6, unitDrunk1)), color3, 0.4);

//
if (length(finalColor) > 1.4)
	finalColor.xy = mix(finalColor.xy, normalizedCoord3, 0.5);
else if (length(finalColor) < 0.4)
	finalColor.yz = mix(finalColor.yz, normalizedCoord3, 0.5);

gl_FragColor = finalColor;
}
);

bool CustomFilter_2::init()
{
return m_program.initWithShaderStrings(g_vshDefaultWithoutTexCoord, s_fsh2);
}

**Comment by developerandroid123** [comment]
Here is the complete filter Can you please check Am i missing something?
Because i am trying to create new twitch filter

**Comment by wysaid** [comment]
You don't know any thing about GLSL.
You should read some tutorial about OpenGL ES2 first, or you won't know what you're doing.

**Comment by developerandroid123** [comment]
yes i need to take some tutorial first and thankyou so much for your time
